### PR TITLE
fix(frontend): Convert PipelineSpec format from json to yaml #7444

### PR DESCRIPTION
**Pull request by jlyaoyuli**
**Description of your changes:**
Updated function to convert backend IR to PipelineSpec type (previous for JSON form; now for Yaml form).

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: frontend/src/data/test/lightweight_python_functions_v2_pipeline_rev.yaml [hunk 1]
```diff
@@ -0,0 +1,236 @@
+pipelineInfo:
```
Comment (zijianjoy) [inline frontend/src/data/test/lightweight_python_functions_v2_pipeline_rev.yaml:1]: Can you also delete the two json files in https://github.com/kubeflow/pipelines/tree/be0d1dad104ca2ed0e543da3013054ba153180fa/frontend/src/data/test, and add the yaml file for `pipeline_with_loops_and_conditions`

File: frontend/src/lib/v2/WorkflowUtils.ts [hunk 2]
```diff
@@ -31,10 +31,17 @@ export function isArgoWorkflowTemplate(template: Workflow): boolean {
...
 
 // Assuming template is the JSON format of PipelineSpec in api/v2alpha1/pipeline_spec.proto
-export function convertJsonToV2PipelineSpec(template: string): PipelineSpec {
-  const pipelineSpecJSON = JSON.parse(template);
-  const ts_pipelinespec = PipelineSpec.fromJSON(pipelineSpecJSON);
-  return ts_pipelinespec;
+export function convertYamlToV2PipelineSpec(template: string): PipelineSpec {
+  // const pipelineSpecJSON = JSON.parse(template);
+  // const ts_pipelinespec = PipelineSpec.fromJSON(pipelineSpecJSON);
+  // return ts_pipelinespec;
```
Comment (zijianjoy) [inline frontend/src/lib/v2/WorkflowUtils.ts:37]: NIT: You can remove these lines.

File: frontend/src/lib/v2/WorkflowUtils.ts [hunk 3]
```diff
@@ -31,10 +31,17 @@ export function isArgoWorkflowTemplate(template: Workflow): boolean {
...
+  const pipelineSpecYAML = jsyaml.safeLoad(template);
+  const ts_pipelinespec = PipelineSpec.fromJSON(pipelineSpecYAML);
+  if (!ts_pipelinespec.root || !ts_pipelinespec.pipelineInfo || !ts_pipelinespec.deploymentSpec) {
+    throw new Error('Important infomation is missing. Pipeline Spec is invalid.');
+  } else {
```
Comment (zijianjoy) [inline frontend/src/lib/v2/WorkflowUtils.ts:42]: You can remove `else` here. That means you can return `ts_pipelinespec` as long as important information is not missing.

File: frontend/src/pages/PipelineDetailsTest.test.tsx [hunk 4]
```diff
@@ -22,7 +22,7 @@ import { ApiExperiment } from 'src/apis/experiment';
 import { ApiPipeline, ApiPipelineVersion } from 'src/apis/pipeline';
 import { ApiRunDetail } from 'src/apis/run';
 import { QUERY_PARAMS, RouteParams } from 'src/components/Router';
-import * as v2PipelineSpec from 'src/data/test/mock_lightweight_python_functions_v2_pipeline.json';
+//import * as v2PipelineSpec from 'src/data/test/mock_lightweight_python_functions_v2_pipeline.json';
```
Comment (zijianjoy) [inline frontend/src/pages/PipelineDetailsTest.test.tsx:25]: Remove this line.

File: frontend/src/pages/PipelineDetailsTest.test.tsx [hunk 5]
```diff
@@ -31,6 +31,12 @@ import * as StaticGraphParser from '../lib/StaticGraphParser';
 import { PageProps } from './Page';
 import PipelineDetails from './PipelineDetails';
 import * as WorkflowUtils from 'src/lib/v2/WorkflowUtils';
+import jsyaml from 'js-yaml';
+import fs from 'fs';
+import YAML from 'yaml';
```
Comment (zijianjoy) [inline frontend/src/pages/PipelineDetailsTest.test.tsx:36]: Do we use this library?

File: frontend/src/pages/PipelineDetailsV2.tsx [hunk 6]
```diff
@@ -122,7 +123,7 @@ function PipelineDetailsV2({
       {selectedTab === 1 && (
         <div className={commonCss.codeEditor} data-testid={'spec-ir'}>
           <Editor
-            value={JSON.stringify(JSON.parse(templateString || ''), null, 2)}
+            value={jsyaml.safeDump(jsyaml.safeLoad(templateString || ''))}
```
Comment (zijianjoy) [inline frontend/src/pages/PipelineDetailsV2.tsx:126]: Why do we load and dump here? It is equivalent to `templateString || ''` directly.

File: frontend/src/pages/PipelineDetailsTest.test.tsx [hunk 7]
```diff
@@ -31,6 +31,12 @@ import * as StaticGraphParser from '../lib/StaticGraphParser';
...
+
+const v2PipelineSpec_path = 'src/data/test/lightweight_python_functions_v2_pipeline_rev.yaml';
```
Comment (zijianjoy) [inline frontend/src/pages/PipelineDetailsTest.test.tsx:38]: Use the following name pattern for constants: `V2_PIPELINESPEC_PATH`

File: frontend/src/pages/PipelineDetailsTest.test.tsx (hunk 4 above)
Comment (jlyaoyuli) [reply frontend/src/pages/PipelineDetailsTest.test.tsx:25]: Ok!

File: frontend/src/lib/v2/WorkflowUtils.ts (hunk 2 above)
Comment (jlyaoyuli) [reply frontend/src/lib/v2/WorkflowUtils.ts:37]: OK!

File: frontend/src/lib/v2/WorkflowUtils.ts (hunk 3 above)
Comment (jlyaoyuli) [reply frontend/src/lib/v2/WorkflowUtils.ts:42]: Oh, I didn't aware this. Thanks for the suggestions! @zijianjoy

File: frontend/src/pages/PipelineDetailsTest.test.tsx (hunk 5 above)
Comment (jlyaoyuli) [reply frontend/src/pages/PipelineDetailsTest.test.tsx:36]: We use fs only. Will delete unused parts.

File: frontend/src/pages/PipelineDetailsV2.tsx (hunk 6 above)
Comment (jlyaoyuli) [reply frontend/src/pages/PipelineDetailsV2.tsx:126]: I just follow the JSON logic to make the changes. You are correct. I can directly use the string! Thank you @zijianjoy !

File: frontend/src/pages/PipelineDetailsTest.test.tsx (hunk 7 above)
Comment (jlyaoyuli) [reply frontend/src/pages/PipelineDetailsTest.test.tsx:38]: OK!

**Comment by zijianjoy** [comment]
/lgtm
/approve

That is awesome work! Thank you Joe!

**Pull request merged**